db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java

```diff
@@ -2790,4 +2794,25 @@ public int getJoinPosition() {
     public CostModel getCostModel() {
         return costModel;
     }
+
+    @Override
+    public void setOuterTableOfJoin(ResultSetNode outerTableOfJoin) {
+        if (outerTableOfJoin == null)
+            this.outerTableOfJoin = null;
+        if (outerTableOfJoin instanceof Optimizable)
+            this.outerTableOfJoin = (Optimizable)outerTableOfJoin;
+    }
+
+    @Override
+    public Optimizable getOuterTable() {
+        if (joinPosition == 0) {
+            if (getJoinType() >= INNERJOIN)
+                return outerTableOfJoin;
+            return null;
+        }
+        int propJoinOrder = proposedJoinOrder[joinPosition-1];
+        Optimizable thisOpt = optimizableList.getOptimizable(propJoinOrder);
+        return thisOpt;
```

Comment:
So, for join position 0, we only return the outer table for OUTER join types, but if position is larger than 0, we return the previous optimizable in the join order, regardless of the join type, is this intentional?

Reply:
This is intentional.  There are two flows to costing a join.  If the syntax 'JOIN ON' is used to join tables and the join cannot be flattened (JoinNode.flatten), then a JoinNode is generated and JoinNode.optimizeSource is called on the right (inner) table.  In this case, the joinPosition will always be zero and joinType reflects if it is an inner join or a type of outer join.  If the join is specified in the FROM clause via comma syntax: FROM t1,t2,t3  ; or if the join is flattened, then SelectNode.findBestPlan is called which iterates through all of the Optimizables in the FROM clause, trying all different combinations.  When joinPosition is zero, we are not planning a join, but a single-table scan.  When joinPosition is greater than 0, then the Optimizable at joinPosition - 1  is the outer table and the Optimizable at joinPosition is the inner table of the join.

Reply:
Great explanation, thank you